pkg/operator/ceph/object/s3-handlers.go

```diff
@@ -90,10 +90,10 @@ func (s *S3Agent) createBucket(name string, infoLogging bool) error {
 			switch aerr.Code() {
 			case s3.ErrCodeBucketAlreadyExists:
 				logger.Debugf("bucket %q already exists", name)
-				return nil
+				return errors.Wrapf(err, "bucket %q already exists", name)
```

Comment:
I believe it's intentional, why would we need to return an error?

Reply:
If I understand correctly I have seen following is the issue for OBC:
when the operator restarts it gets OBC provision request. In Provision()
```go
   err = s3svc.CreateBucket(p.bucketName)
    if err != nil {
        err = errors.Wrapf(err, "error creating bucket %q", p.bucketName)
        logger.Errorf(err.Error())
        p.deleteOBCResourceLogError("")
        return nil, err 
    }
```
here the code expects incase of existing buckets, but here it gets succeeds

Reply:
If it already exists we don't want an error, right? Seems like it already behaves that way.

Reply:
If an error is returned the event (new obc, for example) is re-tried. If the bucket already exists I think we most likely do not want to re-try.

Reply:
@thotz Thoughts on this thread? Don't we already return success if it already exists? I'm still missing why this is needed.

Reply:
The lib bucket provisioner is keeping send ing OBC requests even though creation gets to succeed. And rook operator end ups creating ceph object user on later times and then CreateBucket() returns success this case since AlreadyExists is bypassed. At library creating secrets/cm fails already existing error and this repeated again and again. Above was the orginal issue, @BlaineEXE fix the bug in the [library](https://github.com/kube-object-storage/lib-bucket-provisioner/pull/201) , so I guess we may not end up this scenario.

But there is another negative test case which may need this fix :
* The user have created OBC `abc` with `bucketname: ceph-bkt`. And OBC creation got succeed.
* Now if user tried to create OBC `xyz` with same bucket name, atleast from code reading in library and rook, this request will also succeed but this OBC is won't be useful since actual bucket creation failed for this issue and its credentials is not valid for accessing the bucket. P.S since I don't have set up post rook depend on go v1.15 to test this scenario.

Reply:
If the provisioner returns an error to the lib, then (at least in most cases, AFAIR) the error will cause the OBC event (eg ADD, DELETE, UPDATE) to be re-queued and thus retried. I am not looking at the code at the moment, but if the provisioner returns and error because the passed-in bucket name already exists, then the lib will retry the create (ADD event) until max retries are used. The philosophy for retrying is the user or admin has a chance to correct the problem (eg. update the OBC to use a different storage class or bucket name) and the problem is solved. On the other hand, if the error is not solvable then we end spamming the logs.

Reply:
I think for users we should probably fail if there is a bucket name conflict so that users can't hijack another users' object store by creating a new OBC with a repeat name.

Reply:
Above case is invalid for `Provision()`, but for `Grant()` allows sharing of existing buckets(brownfield case).

Reply:
I do not see support for `Grant` or `Revoke` in the s3-handlers.go -- only `Create` and `Delete`. So, if this is correct, the error case here only applies to creating a new bucket. If I'm wrong and grant calls the same `s.Client.CreateBucket(bucketInput)` then I think you need to implement grant (and revoke) separate from `CreateBucket`.

Reply:
Grant is not calling the create bucket api in the rook provisioner. What I was trying to say to users can give same bucket name for different OBC and OB creation succeeds in this case(negative test case for provisioner) since `already bucket exists` not given from the rook.

```console
# kubectl get obc
NAME                   AGE
ceph-delete-bucket     7m18s
ceph-delete-bucket-1   10s
[root@localhost rook]# kubectl get ob
NAME                               AGE
obc-default-ceph-delete-bucket     4m26s
obc-default-ceph-delete-bucket-1   7s
```
```yaml
$ kubectl describe ob obc-default-ceph-delete-bucket
Name:         obc-default-ceph-delete-bucket
Namespace:    
Labels:       bucket-provisioner=rook-ceph.ceph.rook.io-bucket
Annotations:  <none>
API Version:  objectbucket.io/v1alpha1
Kind:         ObjectBucket
Metadata:
  Creation Timestamp:  2021-04-30T06:48:40Z
  Finalizers:
    objectbucket.io/finalizer
  Generation:        1
  Resource Version:  319087
  Self Link:         /apis/objectbucket.io/v1alpha1/objectbuckets/obc-default-ceph-delete-bucket
  UID:               bbf37800-a4dc-4b1e-a918-bde5441bf7a7
Spec:
  Additional State:
    Ceph User:  ceph-user-j7q0Cb7s
  Claim Ref:
    API Version:  objectbucket.io/v1alpha1
    Kind:         ObjectBucketClaim
    Name:         ceph-delete-bucket
    Namespace:    default
    UID:          ecb69f2f-2df7-44b3-9c1b-d683a951a5ef
  Endpoint:
    Additional Config:
      Max Objects:     60000
      Max Size:        7G
    Bucket Host:       rook-ceph-rgw-my-store.rook-ceph.svc
    Bucket Name:       mybucket1
    Bucket Port:       8081
    Region:            us-east-1
    Sub Region:        
  Reclaim Policy:      Delete
  Storage Class Name:  rook-ceph-delete-bucket
Status:
  Phase:  Bound
Events:   <none>

$ kubectl describe ob obc-default-ceph-delete-bucket-1
Name:         obc-default-ceph-delete-bucket-1
Namespace:    
Labels:       bucket-provisioner=rook-ceph.ceph.rook.io-bucket
Annotations:  <none>
API Version:  objectbucket.io/v1alpha1
Kind:         ObjectBucket
Metadata:
  Creation Timestamp:  2021-04-30T06:52:59Z
  Finalizers:
    objectbucket.io/finalizer
  Generation:        1
  Resource Version:  319665
  Self Link:         /apis/objectbucket.io/v1alpha1/objectbuckets/obc-default-ceph-delete-bucket-1
  UID:               a5b19b5a-30d2-4fa4-9fe2-d374fa5d38ad
Spec:
  Additional State:
    Ceph User:  ceph-user-xEmbGxfu
  Claim Ref:
    API Version:  objectbucket.io/v1alpha1
    Kind:         ObjectBucketClaim
    Name:         ceph-delete-bucket-1
    Namespace:    default
    UID:          1eb6d263-d30f-4077-ba37-d9bb493e416d
  Endpoint:
    Additional Config:
      Max Objects:     80000
      Max Size:        8G
    Bucket Host:       rook-ceph-rgw-my-store.rook-ceph.svc
    Bucket Name:       mybucket1
    Bucket Port:       8081
    Region:            us-east-1
    Sub Region:        
  Reclaim Policy:      Delete
  Storage Class Name:  rook-ceph-delete-bucket
Status:
  Phase:  Bound
Events:   <none>
```
But if ur app uses `ceph-delete-bucket-1` fails since actual bucket creation request failed